src/middleware.rs

```diff
@@ -164,3 +164,81 @@ where
         (self)(request, next)
     }
 }
+
+#[cfg(feature = "digest-auth")]
+pub mod digest {
+    use super::*;
+    use crate::{Request, Response};
+    use digest_auth::{AuthContext, WwwAuthenticateHeader};
+    use std::{borrow::Cow, str::FromStr};
+
+    /// Provides simple digest authentication powered by the `digest_auth` crate.
+    ///
+    /// Requests that receive a HTTP 401 response are retried once by this middleware with the
+    /// credentials provided on construction. The retry only happens under these conditions:
+    /// - there is no prior "authorization" header on the request set by the caller or other
+    ///   middleware, and;
+    /// - the server provides HTTP Digest auth challenge in the "www-authenticate" header.
+    ///
+    /// In other cases, this middleware acts as a no-op forwarder of requests and responses.
+    ///
+    /// ```
+    /// let arbitrary_username = "MyUsername";
+    /// let arbitrary_password = "MyPassword";
+    /// let digest_auth_middleware =
+    ///     ureq::DigestAuthMiddleware::new(arbitrary_username.into(), arbitrary_password.into());
+    /// # let url = String::new();
+    ///
+    /// let agent = ureq::AgentBuilder::new().middleware(digest_auth_middleware).build();
+    /// agent.get(&url).call();
+    /// ```
+    pub struct DigestAuthMiddleware {
+        username: Cow<'static, str>,
+        password: Cow<'static, str>,
+    }
+
+    impl DigestAuthMiddleware {
+        pub fn new(username: Cow<'static, str>, password: Cow<'static, str>) -> Self {
```

Comment:
Sorry for the late comment, but what do you think about taking generics here?

```rust
pub fn new<A: Into<Cow<'static, str>>, B: Into<Cow<'static, str>>>(username: A, password: B) -> Self {
    let username = username.into();
    let password = password.into();
    ...
}
```

Reply:
@strohel actually proposed a very similar thing :D (with `impl Into` instead). I suppose if you two thought the same, it's likely a good idea.

Reply:
I'd definitely prefer to complicate the library functions as opposed to complicating the API - I don't see many public APIs where I have to provide a `Cow` explicitly, so I was surprised for a moment thinking of how I would actually construct one.

Reply:
Sounds good. I'll make this change in the main upstream PR then. (Speaking of which, I'll be closing this PR in favor of that one).